src/main/java/no/digipost/api/useragreements/client/response/StreamingRateLimitedResponse.java

```diff
@@ -0,0 +1,147 @@
+ */
+package no.digipost.api.useragreements.client.response;
+
+import java.time.Instant;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+/**
+ * Represents a streamed response from the Digipost API, and offer mechanisms for consuming
+ * it incrementally. In addition, one should (must) inspect {@link #getNextAllowedRequest()}
+ * after consuming the response and not make any new request against the API resource until
+ * this instant has passed.
+ *
+ * @param <T> The type of the entities contained in the response.
+ */
+public class StreamingRateLimitedResponse<T> {
+
+	@FunctionalInterface
+	public static interface ResponseElementHandler<T> {
+		void handle(T id) throws Exception;
+	}
+
+	private final Stream<T> elements;
+	private final AtomicBoolean consumed = new AtomicBoolean(false);
+	private final Supplier<Instant> nextAllowedRequest;
+
+	public <S extends WithNextAllowedRequestTime> StreamingRateLimitedResponse(Stream<S> responseElements, Function<? super S, Stream<T>> flatMapper) {
+		this(responseElements, new DeferredNextAllowedRequest(), flatMapper);
+	}
+
+	private <S extends WithNextAllowedRequestTime> StreamingRateLimitedResponse(Stream<S> responseElements, DeferredNextAllowedRequest deferredNextAllowedRequest, Function<? super S, Stream<T>> flatMapper) {
+		this(responseElements.flatMap(elements -> {
+			deferredNextAllowedRequest.register(elements);
+			return flatMapper.apply(elements);
+		}), deferredNextAllowedRequest);
+	}
+
+	public StreamingRateLimitedResponse(Stream<T> elements, Supplier<Instant> nextAllowedRequest) {
+		this.elements = elements;
+		this.nextAllowedRequest = nextAllowedRequest;
+	}
+
+	public <R> StreamingRateLimitedResponse<R> map(Function<? super T, R> mapper) {
+		return new StreamingRateLimitedResponse<>(asStream().map(mapper), nextAllowedRequest);
+	}
+
+	public <R> StreamingRateLimitedResponse<R> flatMap(Function<? super T, Stream<R>> mapper) {
+		return new StreamingRateLimitedResponse<>(asStream().flatMap(mapper), nextAllowedRequest);
+	}
+
+	/**
+	 * Consume each element of the response. This will properly terminate the response
+	 * and close associated resources before the method returns, as opposed to {@link #asStream()}
+	 * where you must handle this yourself. When this method returns, it is thus also safe to
+	 * invoke {@link #getNextAllowedRequest()}, as this instant should have been captured from the
+	 * end of the streaming response.
+	 *
+	 * @param handler The function to invoke for each element.
+	 */
+	public void forEach(ResponseElementHandler<T> handler) {
+		try (Stream<T> autoClosed = asStream()) {
+			elements.forEach(id -> {
+				try {
+					handler.handle(id);
+				} catch (RuntimeException e) {
+					throw e;
+				} catch (Exception e) {
+					throw new RuntimeException(e.getClass().getSimpleName() + ": '" + e.getMessage() + "'", e);
```

Comment:
Du kan bruke `ExceptionUtils.exceptionNameAndMessage` som jeg har rappa fra Digg 😎

Reply:
Mente jeg hadde sett noe sånt et sted, men fant det ikke her..? Kan absolutt lage det, men gadd liksom ikke for bare denne.